---
title: "Colibactin clones RNA-Seq analysis - Batch 2"
author: "Hilmar Berger"
output: 
 html_document:
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: true
    code_folding: hide
    
pdf_document:
    fig_caption: true
date: "`r format(Sys.time(), '%d %B, %Y %H:%M:%S')`"
---

```{r, message=FALSE, warning=FALSE}
library(DESeq2)
library(tximport)
library(ggplot2)
library(xlsx)
library(xtable)
library(pheatmap)
library(reshape2)
library(limma)
library(knitr)

result_folder = "../../../../Results/RNASeq"
if (!file.exists(result_folder)) dir.create(result_folder, recursive=T)

```

# Introduction

We here analyze data from an RNASeq experiment (Amina Iftekhar & Michael Sigal) on Wnt-independent clones generated by infection of mouse primary colon epithelial cells with Colibactin-positive E.coli. Details of the experimental procedures used for generating samples have been documented in corresponding EPs. We here study the transcriptome of WNT-independent clones and clones derived from uninfected colon epithelia in identical culture conditions (5 days +Wnt/Rspo followed by 3 days without Wnt/Rspo) for mice 2 and 3.

```{r}
IMPORT_DATA = TRUE

input_folder = "../../../../Data/Processed/RNA_counts/"
salmon_counts_Rdata = file.path(input_folder,"Salmon_gene_expression_batch2.Rdata")

if(IMPORT_DATA) {
  suppressMessages(library(org.Mm.eg.db))
  suppressMessages(library(AnnotationDbi))
  
  tx_anno = read.table("../../../../Data/External/GencodeM12/gencode.vM12.transcript.anno.txt", sep="\t", header=T, stringsAsFactors = F)

  entrez_from_gencode =  read.table(gzfile("../../../../Data/External/GencodeM12/gencode.vM12.metadata.EntrezGene.gz"), sep="\t", header=T, stringsAsFactors = F)
  colnames(entrez_from_gencode) = c("transcript","EntrezID")
  # a few transcripts have more than one Entrez ID - which should not happen in theory. Since most are only predicted proteins we here just ignore them
  dup_tx = entrez_from_gencode[which(duplicated(entrez_from_gencode$transcript)),"transcript"]
  entrez_from_gencode = subset(entrez_from_gencode, !transcript %in% dup_tx)
  tx_anno = merge(tx_anno, unique(entrez_from_gencode), by="transcript", all.x=T, sort=F)
  # fix a missing record that will cause inconsistencies in gene ENSMUSG00000051951
  rownames(tx_anno) = tx_anno$transcript
  tx_anno["ENSMUST00000162897.1",]$EntrezID = 497097
  gene_anno = unique(tx_anno[,c("gene","GeneSymbol", "EntrezID")])
  gene_anno$gene_fixed = unlist(sapply(strsplit(gene_anno$gene,"\\."), function(x) x[1]))
  
  genename=select(org.Mm.eg.db, keys=gene_anno$gene_fixed,         columns=c("GENENAME"),keytype="ENSEMBL") 
  GeneName = as.data.frame(tapply(genename$GENENAME, genename$ENSEMBL, paste, collapse=","))
  colnames(GeneName) = "GeneName"
  gene_anno = merge(gene_anno, GeneName, by.x="gene_fixed", by.y=0, sort=F, all.x=T)
  rownames(gene_anno) = gene_anno$gene
  
  files = list.files(path=input_folder, pattern="3985_.*_quant.sf", full.names = T)
  names(files) = gsub("_quant.sf","",basename(files))
  salmon_gene_expression_data <- tximport(files, type = "salmon", tx2gene = tx_anno[,c("transcript","gene")])
  
  save(salmon_gene_expression_data, tx_anno, gene_anno, file=salmon_counts_Rdata)
} else {
  load(file=salmon_counts_Rdata)
}
```

```{r}
exp_design = read.table("../../../../Data/Metadata/RNA/Batch2/SampleDescription.txt", sep="\t", header=T, stringsAsFactors = F)
rownames(exp_design) = exp_design$sampleID
ed = exp_design
```

# Data overview

## Samples 
```{r,  results='asis'}
ed_tmp = exp_design
row.names(ed_tmp) <- NULL
kable(xtable(ed_tmp,display=rep("s", ncol(ed_tmp)+1), align=paste(paste(rep("|", ncol(ed_tmp)+2), collapse="l"),"|",sep=""), type="html", file="" , include.rownames=F))
```


## Gene abundances and counts

```{r}
boxplot(log10(salmon_gene_expression_data$abundance+1e-5), ylab="log10(Abundance+1e-5)")
boxplot(log10(salmon_gene_expression_data$counts+1e-5), ylab="log10(Counts+1e-5)")

barplot(apply(salmon_gene_expression_data$abundance>0, 2, sum), ylab="# detected genes")
```

```{r}
count_mat = salmon_gene_expression_data$counts
```

## Densities

```{r, densities, fig.width=8, fig.height=8, eval=TRUE}
par(cex=0.8)
cp = rainbow(6)
plotDensities(log2(count_mat+1), legend="topright", group=ed[colnames(count_mat),"MouseID"], col=cp)
```

## Total mapped read numbers

Pleae note that those counts are based on estimated counts from the combined transcript abundances. There might therefore be small differences to absolute read counts (mapped to transcriptome) per sample. 

```{r, readnum, fig.width=8, fig.height=8, eval=TRUE}
par(mar=c(5,10,4,2))
barplot(apply(count_mat,2,sum),las=2, ylab="Log10(Total mapped read number)", main="Total read number")
apply(count_mat,2,sum)
```


## Sample correlation

```{r, cor_mat, fig.width=8, fig.height=8}
sample_ordered = rownames(exp_design)
cor_mat = cor(log10(count_mat+1), method="spearman")
cor_mat = cor_mat[sample_ordered, sample_ordered]
# pheatmap(cor_mat, scale="none")

cor_mat2 = cor_mat
rownames(cor_mat2) = paste(ed[rownames(cor_mat2),]$SampleName, rownames(cor_mat2),sep=" - ")
colnames(cor_mat2) = ed[colnames(cor_mat2),]$SampleName
pheatmap(cor_mat2,scale="none", cluster_rows=F, cluster_cols=F )

```

## Multi Dimensional Scaling on all genes

```{r, MDS_all, fig.width=8, fig.height=8}
cp = palette(rainbow(3))
data_inp = t(log10(count_mat+1)) 

d <- dist(data_inp) # euclidean distances between the rows
fit <- cmdscale(d,eig=TRUE, k=2) # k is the number of dim

# plot solution
x <- fit$points[,1]
y <- fit$points[,2]
plot(x, y, xlab="Coordinate 1", ylab="Coordinate 2", main="Metric MDS, all samples", type="n")
text(x,y,labels=ed[rownames(data_inp),]$SampleName, col=cp[as.numeric(factor(ed[rownames(data_inp),]$Condition))])
```




# Differential Expression Analysis

We compare WNT-independent vs uninfected control conditions for all WI clones combined and for individidual WI clones.

```{r}
coldata <- exp_design[colnames(salmon_gene_expression_data$counts),]
rownames(coldata) <- coldata$sampleID
ddsSalmonGE <- DESeqDataSetFromTximport(salmon_gene_expression_data, colData=coldata, design=~ Condition)
```

```{r, dge_analysis, results='hide', warning=FALSE, message=FALSE}
count_mat = counts(ddsSalmonGE)
ed = exp_design

included_genes = (apply(count_mat,1,max) > 1) # & (substr(rownames(count_mat),1,1)!="_")

all_results = list()

############################################
# All samples
############################################

included_samples = rownames(ed)

dds_selected = ddsSalmonGE[included_genes, included_samples]
dds_selected$Condition <- factor(dds_selected$Condition, levels=c("WI","NI"))
dds_selected <- DESeq(dds_selected, fitType="local")

res <- results(dds_selected, contrast=c("Condition","WI","NI"))

all_results[["WI_vs_uninfected_all_samples"]] = res


############################################
# Mouse 2 WI vs NI
############################################

included_samples = rownames(subset(ed,  MouseID == "M2"))

dds_selected = ddsSalmonGE[included_genes, included_samples]
dds_selected$Condition <- factor(dds_selected$Condition, levels=c("WI","NI"))
dds_selected <- DESeq(dds_selected, fitType="local")

res <- results(dds_selected, contrast=c("Condition","WI","NI"))

all_results[["M2_WI_vs_uninfected"]] = res

############################################
# Mouse 3 WI vs NI
############################################

included_samples = rownames(subset(ed,  MouseID == "M3"))

dds_selected = ddsSalmonGE[included_genes, included_samples]
dds_selected$Condition <- factor(dds_selected$Condition, levels=c("WI","NI"))
dds_selected <- DESeq(dds_selected, fitType="local")

res <- results(dds_selected, contrast=c("Condition","WI","NI"))

all_results[["M3_WI_vs_uninfected"]] = res

############################################################################################

excluded_genes = names(included_genes[!included_genes])
cols_in_res_table = colnames(all_results[[1]])
empty_cols = do.call(cbind, Map(function(x) rep(NA, length(excluded_genes)), cols_in_res_table ))
excluded_gene_tab = data.frame(row.names = excluded_genes, empty_cols )

for (n in names(all_results)) {
  tmp = all_results[[n]]
  resOrdered <- data.frame(tmp[order(tmp$pvalue),])
  resOrdered = rbind(resOrdered, excluded_gene_tab)
  resOrdered$padj = ifelse(is.na(resOrdered$padj),1,resOrdered$padj)
  res_annotated = as.data.frame(merge(resOrdered, gene_anno[,c("gene","GeneSymbol","EntrezID","GeneName")], by.x=0, by.y="gene", sort=F))
  res_annotated = merge(res_annotated, count_mat, by.x="Row.names",by.y=0, all.x=T, sort=F)
  all_results[[n]] = res_annotated
}

```

### Volcano plots

For each comparison, the distribution of the fold change (on a log2 scale) and adjusted p-value (on reverse logarithmical scale) is shown in a volcano plot. The red line denotes the FDR cutoff of 0.05. 

```{r, DE_volcano, results="hide", fig.width=16,fig.height = 10 }
all_target_conditions = names(all_results)
par(mfrow=c(2,2))
all_GSA_results = list()
for (tc in all_target_conditions) {
  r = all_results[[tc]]
  plot(r$log2FoldChange, -log10(r$padj),xlab="log2 Fold Change",ylab="-log10(adj. p-val)", ylim=c(0,max(2,max(-log10(r$padj),na.rm=T))))
  title(main=tc, sub=paste("(",nrow(subset(r, padj < 0.05))," signif. DE genes)",sep="") )
  abline(h=-log10(0.05),col="red")
  
}
```



```{r, DE_combined, results="hide"}
all_DE_results_tmp = list()
for (tc in all_target_conditions) {
  tmp = all_results[[tc]]
  tmp$condition = tc
  all_DE_results_tmp[[tc]] = tmp
}
all_DE_results_ts = do.call(rbind, all_DE_results_tmp)
all_DE_results_ts$DE_class = ifelse(all_DE_results_ts$padj>0.05, "n.s.", ifelse(all_DE_results_ts$log2FoldChange > 0,"Up","Down"))
#agg_fun = function(x) paste(unique(x),collapse=";")
agg_fun = function(x) ifelse("Down" %in% x, "Down",ifelse("Up" %in% x, "Up","n.s."))
all_DE_results_sw = dcast(all_DE_results_ts, GeneSymbol ~ condition, value.var="DE_class", fun.aggregate=agg_fun)
```


```{r, message=FALSE, warning=FALSE, results='hide'}
all_output_txt_files = list()
all_output_excel_files = list()
output_file_prefix = paste(result_folder,"Differential_expression_results_batch2_", sep="/")
#selected_cols = c("Row.names", "GeneSymbol", "GeneDescription","logFC","AveExpr","t","P.Value","adj.P.Val" )
selected_cols = colnames(all_results[[1]])
for (tc in all_target_conditions) {
  filename = paste(output_file_prefix, tc, ".txt", sep="" )
  write.table(all_results[[tc]][,selected_cols], file= filename, row.names=F , sep="\t", dec=".", quote=F)
  all_output_txt_files[[paste("DGE",tc)]] = filename
}


filename = paste(result_folder,"DE_results_comparison_batch2.txt",sep="/")
write.table(all_DE_results_sw, file = filename,sep="\t",quote=F, row.names=F)
all_output_txt_files[["DGE comparison"]] = filename

tmp = as.data.frame(assay(varianceStabilizingTransformation(ddsSalmonGE, blind=FALSE)))
#colnames(tmp) = ed[colnames(tmp),]$Sample
#rownames(tmp) = normalized$genes$ProbeName
tmp = merge(tmp, gene_anno, by.x=0, by.y="gene", all.x=T, sort=F)
filename = paste(result_folder,"Normalized_expression_data_batch2.txt",sep="/")
write.table(tmp, file=filename,sep="\t",col.names=NA)

all_output_txt_files[["Normalized expression values"]] = filename

```


# Result files

The following files have been written:
```{r, echo=FALSE}
output_file = paste(result_folder,"DGE_analysis_image_batch2.Rdata", sep="/")
save(all_results, count_mat, ed, gene_anno, file=output_file)

all_txt_files = as.data.frame(t(as.data.frame(all_output_txt_files)))
colnames(all_txt_files)[1] = "File name"
all_txt_files$Format = "Tab separated text"

all_txt_files

```


```{r}
sessionInfo()
```

